core/src/main/java/com/linecorp/armeria/client/Http2ResponseDecoder.java

```diff
@@ -208,6 +210,12 @@ public void onHeadersRead(ChannelHandlerContext ctx, int streamId, Http2Headers
 
         if (endOfStream) {
             res.close();
+
+            if (needsToDisconnectNow() && !goAwayHandler.sentGoAway() && !goAwayHandler.receivedGoAway()) {
```

Comment:
How about extracting this into a boolean-returning method so we can deduplicate the `if` below?